crates/hir-def/src/item_tree.rs

```diff
@@ -115,20 +115,36 @@ impl ItemTree {
             return Default::default();
         }
 
-        let ctx = lower::Ctx::new(db, file_id);
         let mut top_attrs = None;
         let mut item_tree = match_ast! {
             match syntax {
                 ast::SourceFile(file) => {
+                    let ctx = lower::Ctx::new(db, file_id);
                     top_attrs = Some(RawAttrs::new(db, &file, ctx.hygiene()));
                     ctx.lower_module_items(&file)
                 },
                 ast::MacroItems(items) => {
+                    // If a fn-like macro is defined in a trait, it should have the inherited visibility
+                    let trait_node = {
+                        if let Some((_, original_node)) = file_id.original_call_node(db.upcast()) {
+                            original_node.ancestors().filter_map(|node| syntax::ast::Trait::cast(node)).next()
```

Comment:
Using ancestors as below won't properly work, as we could missclassify if we are a local item in an associated item of a trait. That is

```rs
macro_rules! call {
    () => {fn func() {}};
}
trait T {
    fn func() {
        call!()
    }
}